packages/api/src/routers/api/alerts.ts

```diff
@@ -1,210 +1,110 @@
-  AlertChannel,
-  AlertInterval,
-  AlertType,
-  AlertSource,
-} from '../../models/alert';
-import * as clickhouse from '../../clickhouse';
-import { SQLSerializer } from '../../clickhouse/searchQueryParser';
+import Alert from '../../models/alert';
 import { getTeam } from '../../controllers/team';
 import { isUserAuthenticated } from '../../middleware/auth';
+import {
+  createAlert,
+  updateAlert,
+  validateGroupByProperty,
+} from '../../controllers/alerts';
 
 const router = express.Router();
 
-const getCron = (interval: AlertInterval) => {
-  const now = new Date();
-  const nowMins = getMinutes(now);
-  const nowHours = getHours(now);
-
-  switch (interval) {
-    case '1m':
-      return '* * * * *';
-    case '5m':
-      return '*/5 * * * *';
-    case '15m':
-      return '*/15 * * * *';
-    case '30m':
-      return '*/30 * * * *';
-    case '1h':
-      return `${nowMins} * * * *`;
-    case '6h':
-      return `${nowMins} */6 * * *`;
-    case '12h':
-      return `${nowMins} */12 * * *`;
-    case '1d':
-      return `${nowMins} ${nowHours} * * *`;
-  }
-};
+// Input validation
+const zChannel = z.object({
+  type: z.literal('webhook'),
+  webhookId: z.string().min(1),
+});
 
-const createAlert = async ({
-  channel,
-  groupBy,
-  interval,
-  logViewId,
-  threshold,
-  type,
-}: {
-  channel: AlertChannel;
-  groupBy?: string;
-  interval: AlertInterval;
-  logViewId: string;
-  threshold: number;
-  type: AlertType;
-}) => {
-  return new Alert({
-    channel,
-    cron: getCron(interval),
-    groupBy,
-    interval,
-    source: AlertSource.LOG,
-    logView: logViewId,
-    threshold,
-    timezone: 'UTC', // TODO: support different timezone
-    type,
-  }).save();
-};
+const zLogAlert = z.object({
+  source: z.literal('LOG'),
+  groupBy: z.string().optional(),
+  logViewId: z.string().min(1),
+  message: z.string().optional(),
+});
 
-// create an update alert function based off of the above create alert function
-const updateAlert = async ({
-  channel,
-  groupBy,
-  id,
-  interval,
-  logViewId,
-  threshold,
-  type,
-}: {
-  channel: AlertChannel;
-  groupBy?: string;
-  id: string;
-  interval: AlertInterval;
-  logViewId: string;
-  threshold: number;
-  type: AlertType;
-}) => {
-  return Alert.findByIdAndUpdate(
-    id,
-    {
-      channel,
-      cron: getCron(interval),
-      groupBy: groupBy ?? null,
-      interval,
-      source: AlertSource.LOG,
-      logView: logViewId,
-      threshold,
-      timezone: 'UTC', // TODO: support different timezone
-      type,
-    },
-    {
-      returnDocument: 'after',
-    },
-  );
-};
+const zChartAlert = z.object({
+  source: z.literal('CHART'),
+  chartId: z.string().min(1),
+  dashboardId: z.string().min(1),
+});
 
-router.post('/', isUserAuthenticated, async (req, res, next) => {
-  try {
+const zAlert = z
+  .object({
+    channel: zChannel,
+    interval: z.enum(['1m', '5m', '15m', '30m', '1h', '6h', '12h', '1d']),
+    threshold: z.number().min(1),
```

Comment:
prev webapp version sent a `string` to `threshold`.
@MikeShi42 does it have to be backwards-compatible if `api` and `app` versions differ in deployment? or should they always be in sync?

Reply:
They're generally deployed at the same time, in general the app will roll out before the API so it should be safe given the backend I believe will have the stricter validation after the app is rolled out. Though we can also time the deployment pipeline manually if needed.

Reply:
good to know